File: civic_pulse_frontend/lib/models/complaint_report.dart
import 'dart:convert';

import 'package:civic_pulse_frontend/models/comments.dart';
import 'package:civic_pulse_frontend/models/feedback.dart';
import 'package:civic_pulse_frontend/models/history.dart';

class ComplaintReport {
  final String id;
  final String image;
  final String userId;
  final String location;
  final String desc;
  final DateTime createdAt;
  final String profilePic;
  final bool escalated;
  final Map<String, dynamic> assignedHead;
  final List<Map<String, dynamic>> assignedTo;
  final List<History> history;
  final String status;
  final int upvote;
  final String phone;
  final List<Comments> comments;
  final String department;
  final String fullname;
  final String? completionImage;
  final Feedback? feedback;
  final DateTime updatedAt;

  ComplaintReport({
    required this.id,
    required this.image,
    required this.location,
    required this.desc,
    required this.phone,
    required this.department,
    required this.fullname,
    required this.userId,
    required this.status,
    required this.escalated,
    required this.profilePic,
    required this.assignedHead,
    required this.assignedTo,
    required this.history,
    this.upvote = 0,
    this.comments = const [],
    this.completionImage,
    this.feedback,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ComplaintReport.fromMap(Map<String, dynamic> map) {
    // Safe upvote parsing
    int parsedUpvote = 0;
    if (map['upvote'] != null) {
      if (map['upvote'] is int) {
        parsedUpvote = map['upvote'];
      } else {
        parsedUpvote = int.tryParse(map['upvote'].toString()) ?? 0;
      }
    }

    return ComplaintReport(
      id: map['_id']?.toString() ?? '',
      image: map['image']?.toString() ?? '',
      location: map['location']?.toString() ?? '',
      desc: map['desc']?.toString() ?? '',
      userId: map['userId']?.toString() ?? '',
      profilePic: map['profilePic']?.toString() ?? '',
      escalated: map['escalated'] ?? false,
      assignedHead: map['assignedHead'] is Map
          ? Map<String, dynamic>.from(map['assignedHead'])
          : {},
      assignedTo: map['assignedTo'] is List
          ? List<Map<String, dynamic>>.from(
          map['assignedTo'].map((x) => Map<String, dynamic>.from(x)))
          : [],
      history: map['history'] is List
          ? List<History>.from(map['history'].map((x) => History.fromMap(x)))
          : [],
      status: map['status']?.toString() ?? '',
      upvote: parsedUpvote,
      phone: map['phone']?.toString() ?? '',
      comments: map['comments'] is List
          ? List<Comments>.from(
          map['comments'].map((x) => Comments.fromMap(x)))
          : [],
      department: map['department']?.toString() ?? '',
      fullname: map['fullname']?.toString() ?? '',
      completionImage: map['completionImage']?.toString(),
      feedback: map['feedback'] != null
          ? Feedback.fromMap(Map<String, dynamic>.from(map['feedback']))
          : null,
      createdAt: map['createdAt'] != null
          ? DateTime.tryParse(map['createdAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
      updatedAt: map['updatedAt'] != null
          ? DateTime.tryParse(map['updatedAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'image': image,
      'location': location,
      'desc': desc,
      'phone': phone,
      'profilePic': profilePic,
      'assignedHead': assignedHead,
      'escalated': escalated,
      'assignedTo': assignedTo,
      'history': history.map((x) => x.toMap()).toList(),
      'status': status,
      'upvote': upvote,
      'comments': comments.map((x) => x.toMap()).toList(),
      'department': department,
      'fullname': fullname,
      'completionImage': completionImage,
      'feedback': feedback?.toMap(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'userId': userId,
    };
  }

  factory ComplaintReport.fromJson(String source) =>
      ComplaintReport.fromMap(json.decode(source));

  String toJson() => json.encode(toMap());
}

File: civic_pulse_backend/models/complaint_report.js
const mongoose = require('mongoose');

const commentSchema = new mongoose.Schema({
  userId: { type: String, required: true },
  fullname: { type: String, required: true },
  profilePic: { type: String },
  text: { type: String, required: true },
}, { timestamps: true });

// Feedback schema (from citizen after resolution)
const feedbackSchema = new mongoose.Schema({
  rating: { type: Number, min: 1, max: 5 }, // optional star rating
  comment: { type: String },
  submittedBy: { type: String, required: true }, // citizen userId
}, { timestamps: true });

// Tracking history (status updates with timestamps)
const historySchema = new mongoose.Schema({
  status: { type: String, enum: ['pending', 'processing', 'completed', 'reopened'], required: true },
  updatedAt: { type: Date, default: Date.now },
  updatedBy: { type: String }, // workerId or headId
  note: { type: String }, // e.g., "Assigned to Worker A"
});

const complainReportSchema = new mongoose.Schema({
  image: { type: String }, // complaint image
  userId: { type: String, required: true }, // citizen who created
  location: { type: String, required: true },
  status: { type: String, enum: ['pending', 'processing', 'completed', 'reopened'], default: 'pending' },
  department: { type: String, required: true },
  upvote: { type: Number, default: 0 },
  profilePic: { type: String },
  desc: { type: String, required: true },
  phone: { type: String, required: true },
  fullname: { type: String, required: true },
  assignedHead:{
    headId: { type: String },
    headName: { type: String },
  },
  assignedTo: [
    {
      workerId: { type: String },
      workerName: { type: String },
    }
  ],
  completionImage: { type: String }, // after fix photo
  escalated: { type: Boolean, default: false },
  feedback: feedbackSchema, // citizen feedback
  history: [historySchema], // status timeline
  comments: [commentSchema],
}, { timestamps: true });

const ComplainReport = mongoose.model('ComplaintReport', complainReportSchema);
module.exports = ComplainReport;
